### PR TITLE
[develop] Trello-7IByLDvd: Get currency rate

### PR DESCRIPTION
**Pull request by AndreySurzhan**
[Trello-7IByLDvd](https://trello.com/c/7IByLDvd): Get currency rate

GET `'/currencies/exchange-rate/:serviceId'`
GET `/currencies/exchange-rate'`

### CODE REVIEW

File: controllers/exchangerateservice.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+        return deferred.promise;
+    },
+
+    getCurrencyExchangeRate: function (data) {
+        var deferred = Q.defer();
+
+        this.getById(data.id)
+            .then(function (exchangeRateService) {
+                console.log(isUpdateRequired(exchangeRateService))
```
Comment (AndreySurzhan) [inline controllers/exchangerateservice.js]: remove

File: package.json [hunk 2]
```diff
@@ -1,25 +1,28 @@
...
-        "oauth2orize": "^1.2.2",
-        "passport": "^0.3.2",
-        "passport-http": "^0.3.0",
-        "passport-http-bearer": "^1.0.1",
-        "passport-oauth2-client-password": "^0.1.2",
-        "q": "1.4.1",
-        "underscore": "^1.8.3",
-        "winston": "^2.2.0"
-    }
+  "name": "mk-api",
```
Comment (AndreySurzhan) [inline package.json]: indent

File: controllers/exchangerateservice.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+            return rate;
+        }
+
+        function getECBRatesObject (response) {
+            var currency;
+            var doc = new DOMParser().parseFromString(response);
+            var rate = {};
+            var i;
+            var rateValue;
```
Comment (AndreySurzhan) [inline controllers/exchangerateservice.js]: sort

File: controllers/exchangerateservice.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                    return null;
+            }
+        }
+
+        function getCBRRatesObject (response) {
+            var charCode;
+            var doc = new DOMParser().parseFromString(response);
+            var rate = {};
+            var i;
+            var rateValue;
```
Comment (AndreySurzhan) [inline controllers/exchangerateservice.js]: sort

**Comment by AndreySurzhan** [comment]
> 
{
    "_id": "57a8ea4f9179852398609bf7",
    "name": "Центральный банк Российской Федерации",
    "abbreviation": "CBR",
    "url": "http://www.cbr.ru/scripts/XML_daily_eng.asp",
    "lastUpdate": "2016-08-20T05:09:58.161Z",
    "updateTime": "14:00:00 GMT+0300",
    "base": "RUB",
    "rates": {
      "AUD": "48,6129",
      ----------------------
      "JPY": "63,8147"
    },
    "created": "2016-08-08T20:23:42.143Z"
  },
  {
    "_id": "57a8ea4f9179852398609bf8",
    "name": "European Central Bank",
    "abbreviation": "ECB",
    "url": "http://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml",
    "lastUpdate": "2016-08-20T05:11:52.906Z",
    "updateTime": "16:00:00 GMT+0100",
    "base": "EUR",
    "rates": {
      "USD": "1.1326",
      -----------------
      "ZAR": "15.3019"
    },
    "created": "2016-08-08T20:23:42.143Z"
  },
  {
    "_id": "57a8ea4f9179852398609bf9",
    "name": "Open Exchange Rates API (forex)",
    "abbreviation": "OER",
    "url": "https://openexchangerates.org/api/latest.json?app_id=f30c1cbbe48445d3af6a995d7fd9f30b",
    "lastUpdate": "2016-08-20T05:12:45.178Z",
    "updatePeriod": 3600,
    "base": "USD",
    "rates": {
      "AED": 3.672929,
      -----------------
      "ZWL": 322.322775
    },
    "created": "2016-08-08T20:23:42.143Z"
>

**Comment by AndreySurzhan** [comment]
@wildeastengineer review please, PR is already rebased

File: controllers/exchangerateservice.js [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+     */
+    getCurrencyExchangeRate: function (id) {
+        var deferred = Q.defer();
+
+        this.getById(id)
+            .then(function (exchangeRateService) {
+                if (!isUpdateRequired(exchangeRateService)){
+                    deferred.resolve(exchangeRateService);
+
+                    return;
```
Comment (wildeastengineer) [inline controllers/exchangerateservice.js]: `return deferred.promise;` ?

File: controllers/exchangerateservice.js [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+                }
+
+                return requestCurrencyRates(exchangeRateService.url)
```
Comment (wildeastengineer) [inline controllers/exchangerateservice.js]: I think `return` is extra

File: controllers/exchangerateservice.js [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+                    .then(function (response) {
```
Comment (wildeastengineer) [inline controllers/exchangerateservice.js]: Can you make more flat? Like this
```
f1().then(function () {
    /* --- */
}).then(function () {
    /* --- */
}).then(function () {
    /* --- */
})
```

File: controllers/exchangerateservice.js [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+            return deferred.promise;
+        }
+
+        /**
+         * Get currency exchange rate object
+         *
+         * @param {string} serviceName
+         * @param {string | JSON} response
+         *
+         * @returns {Object | null} rate
```
Comment (wildeastengineer) [inline controllers/exchangerateservice.js]: `{?Object}` [Nullable type](http://usejsdoc.org/tags-type.html)

File: controllers/exchangerateservice.js [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+                if (!(charCode.length === 1 && rateValue.length === 1)) {
+                    continue;
+                }
+
+                charCode = charCode[0];
+                rateValue = rateValue[0];
+
+                if (charCode.hasChildNodes() && rateValue.hasChildNodes()
+                    && charCode.firstChild.nodeValue === charCode.lastChild.nodeValue
```
Comment (wildeastengineer) [inline controllers/exchangerateservice.js]: padding should be 8 spaces

File: controllers/exchangerateservice.js [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+                    && rateValue.firstChild.nodeValue === rateValue.lastChild.nodeValue){
```
Comment (wildeastengineer) [inline controllers/exchangerateservice.js]: And I thought that `&&` should be in the end of line but I like variant more :smile:

**Pull request merged**